### PR TITLE
Ag/refactoring/clean architecture pattern

### PR DESCRIPTION
**Pull request by andriy1507**
Please, check is it correct or not.

### CODE REVIEW

File: remote/src/main/java/com/goryachok/remote/model/RemoteEntity.kt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package com.goryachok.remote.model
+
+import com.google.gson.Gson
+import java.lang.reflect.Type
+
+abstract class RemoteEntity : Type {
```
Comment (EugeneKryvoshey) [inline remote/src/main/java/com/goryachok/remote/model/RemoteEntity.kt]: Please, use type 'interface' and remove inheritance.

File: remote/src/main/java/com/goryachok/remote/model/Main.kt [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package com.goryachok.remote.model
+
+
+import android.os.Parcelable
+import com.google.gson.annotations.SerializedName
+import kotlinx.android.parcel.Parcelize
+
+@Parcelize
+data class Main(
```
Comment (EugeneKryvoshey) [inline remote/src/main/java/com/goryachok/remote/model/Main.kt]: Please inherit this class from 'RemoteEntity'.

File: remote/src/main/java/com/goryachok/remote/model/Forecast.kt [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+import android.os.Parcelable
+import com.google.gson.annotations.SerializedName
+import com.goryachok.remote.model.Main
+import com.goryachok.remote.model.Weather
+import com.goryachok.remote.model.Wind
+import kotlinx.android.parcel.Parcelize
+
+@Parcelize
+data class Forecast(
```
Comment (EugeneKryvoshey) [inline remote/src/main/java/com/goryachok/remote/model/Forecast.kt]: Please inherit this class from 'RemoteEntity'.

File: remote/src/main/java/com/goryachok/remote/model/Coordinates.kt [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package com.goryachok.remote.model
+
+
+import android.os.Parcelable
+import com.google.gson.annotations.SerializedName
+import kotlinx.android.parcel.Parcelize
+
+@Parcelize
+data class Coordinates(
```
Comment (EugeneKryvoshey) [inline remote/src/main/java/com/goryachok/remote/model/Coordinates.kt]: Please inherit this class from 'RemoteEntity'.

File: remote/src/main/java/com/goryachok/remote/model/City.kt [hunk 5]
```diff
@@ -0,0 +1,24 @@
+package com.goryachok.remote.model
+
+
+import android.os.Parcelable
+import com.google.gson.annotations.SerializedName
+import kotlinx.android.parcel.Parcelize
+
+@Parcelize
+data class City(
```
Comment (EugeneKryvoshey) [inline remote/src/main/java/com/goryachok/remote/model/City.kt]: Please inherit this class from 'RemoteEntity'.

File: local/src/main/java/com/goryachok/local/LocalDataSource.kt [hunk 6]
```diff
@@ -0,0 +1,20 @@
+package com.goryachok.local
+
+import android.content.SharedPreferences
+import com.goryachok.local.model.LocalForecast
+import com.goryachok.local.model.LocalWeather
+
+interface LocalDataSource {
+
+    val preferences:SharedPreferences
```
Comment (EugeneKryvoshey) [inline local/src/main/java/com/goryachok/local/LocalDataSource.kt]: Please, don't use public fields in interface unless you have to retrieve this data from implemented object.

File: repository/src/main/java/com/goryachok/repository/WeatherRepository.kt [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+import com.goryachok.local.LocalDataSource
+import com.goryachok.local.model.LocalWeather
+import com.goryachok.remote.RemoteDataSource
+import com.goryachok.remote.model.ApiResponse
+import com.goryachok.remote.model.WeatherEntity
+
+
+interface WeatherRepository {
+
+    val local: LocalDataSource
```
Comment (EugeneKryvoshey) [inline repository/src/main/java/com/goryachok/repository/WeatherRepository.kt]: Please remove this.

File: repository/src/main/java/com/goryachok/repository/WeatherRepository.kt [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+    val remote: RemoteDataSource
```
Comment (EugeneKryvoshey) [inline repository/src/main/java/com/goryachok/repository/WeatherRepository.kt]: Please remove this.

File: repository/src/main/java/com/goryachok/repository/ForecastRepository.kt [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+import com.goryachok.local.LocalDataSource
+import com.goryachok.local.model.LocalForecast
+import com.goryachok.remote.RemoteDataSource
+import com.goryachok.remote.model.ForecastEntity
+import com.goryachok.remote.model.ApiResponse
+
+interface ForecastRepository {
+
+    val local: LocalDataSource
```
Comment (EugeneKryvoshey) [inline repository/src/main/java/com/goryachok/repository/ForecastRepository.kt]: Please remove this.

File: repository/src/main/java/com/goryachok/repository/ForecastRepository.kt [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+    val remote: RemoteDataSource
```
Comment (EugeneKryvoshey) [inline repository/src/main/java/com/goryachok/repository/ForecastRepository.kt]: Please remove this.

File: repository/src/main/java/com/goryachok/repository/WeatherRepository.kt [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+    suspend fun getRemoteDataByCity(city: String): ApiResponse<WeatherEntity>
+
+    fun getLocalData(): LocalWeather
+
+    suspend fun getDataByCoordinates(lat: Float, lon: Float): ResponseResult<LocalWeather>
+
+    suspend fun getDataByCity(city: String): ResponseResult<LocalWeather>
+
+    fun isFetchNeeded(): Boolean
```
Comment (EugeneKryvoshey) [inline repository/src/main/java/com/goryachok/repository/WeatherRepository.kt]: Please remove this.

File: repository/src/main/java/com/goryachok/repository/ForecastRepository.kt [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+    suspend fun getRemoteDataByCity(city: String): ApiResponse<ForecastEntity>
+
+    fun getLocalData(): LocalForecast
+
+    suspend fun getDataByCoordinates(lat: Float, lon: Float): ResponseResult<LocalForecast>
+
+    suspend fun getDataByCity(city: String): ResponseResult<LocalForecast>
+
+    fun isFetchNeeded(): Boolean
```
Comment (EugeneKryvoshey) [inline repository/src/main/java/com/goryachok/repository/ForecastRepository.kt]: Please remove this.

File: remote/src/main/java/com/goryachok/remote/model/Wind.kt [hunk 13]
```diff
@@ -0,0 +1,14 @@
+package com.goryachok.remote.model
+
+
+import android.os.Parcelable
+import com.google.gson.annotations.SerializedName
+import kotlinx.android.parcel.Parcelize
+
+@Parcelize
+data class Wind(
```
Comment (EugeneKryvoshey) [inline remote/src/main/java/com/goryachok/remote/model/Wind.kt]: Please inherit this class from 'RemoteEntity' interface.

File: remote/src/main/java/com/goryachok/remote/model/Weather.kt [hunk 14]
```diff
@@ -0,0 +1,12 @@
+package com.goryachok.remote.model
+
+
+import android.os.Parcelable
+import com.google.gson.annotations.SerializedName
+import kotlinx.android.parcel.Parcelize
+
+@Parcelize
+data class Weather(
```
Comment (EugeneKryvoshey) [inline remote/src/main/java/com/goryachok/remote/model/Weather.kt]: Please inherit this class from 'RemoteEntity' interface.

File: remote/src/main/java/com/goryachok/remote/RemoteDataSource.kt [hunk 15]
```diff
@@ -0,0 +1,18 @@
+package com.goryachok.remote
+
+import com.goryachok.remote.model.ForecastEntity
+import com.goryachok.remote.model.WeatherEntity
+import com.goryachok.remote.model.ApiResponse
+
+interface RemoteDataSource {
+
+    val api: WeatherApiService
```
Comment (EugeneKryvoshey) [inline remote/src/main/java/com/goryachok/remote/RemoteDataSource.kt]: Please remove this.

File: local/src/main/java/com/goryachok/local/model/LocalWeather.kt [hunk 16]
```diff
@@ -0,0 +1,22 @@
+package com.goryachok.local.model
+
+import android.os.Parcelable
+import com.google.gson.Gson
+import kotlinx.android.parcel.Parcelize
+
+@Parcelize
+data class LocalWeather(
```
Comment (EugeneKryvoshey) [inline local/src/main/java/com/goryachok/local/model/LocalWeather.kt]: Please, rename to WeatherLocal.

File: local/src/main/java/com/goryachok/local/model/LocalData.kt [hunk 17]
```diff
@@ -0,0 +1,3 @@
+package com.goryachok.local.model
+
+interface LocalData
```
Comment (EugeneKryvoshey) [inline local/src/main/java/com/goryachok/local/model/LocalData.kt]: Please rename to LocalEntity

File: local/src/main/java/com/goryachok/local/LocalDataSource.kt [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+
+    fun saveForecastData(data: LocalForecast)
+
+    fun saveWeatherData(data: LocalWeather)
+
+    fun readForecastData(): LocalForecast
+
+    fun readWeatherData(): LocalWeather
+
+    fun isDataAvailable(): Boolean
```
Comment (EugeneKryvoshey) [inline local/src/main/java/com/goryachok/local/LocalDataSource.kt:16]: Please remove this.

**Review by EugeneKryvoshey — CHANGES_REQUESTED** [review]

File: repository/src/main/java/com/goryachok/repository/di/RepositoryComponent.kt [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+import com.goryachok.core.RemoteDataSourceProvider
+import com.goryachok.core.RepositoryProvider
+import com.goryachok.repository.ForecastRepository
+import com.goryachok.repository.WeatherRepository
+import dagger.Component
+
+@Component(dependencies = [LocalDataSourceProvider::class, RemoteDataSourceProvider::class])
+interface RepositoryComponent:RepositoryProvider {
+
+    fun inject(repo: ForecastRepository)
```
Comment (EugeneKryvoshey) [inline repository/src/main/java/com/goryachok/repository/di/RepositoryComponent.kt]: Please remove this.

File: repository/src/main/java/com/goryachok/repository/di/RepositoryComponent.kt [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+
+    fun inject(repo: WeatherRepository)
```
Comment (EugeneKryvoshey) [inline repository/src/main/java/com/goryachok/repository/di/RepositoryComponent.kt]: Please remove this.

File: remote/src/main/java/com/goryachok/remote/model/RemoteEntity.kt (hunk 1 above)
Comment (andriy1507) [reply remote/src/main/java/com/goryachok/remote/model/RemoteEntity.kt]: Done

File: local/src/main/java/com/goryachok/local/LocalDataSource.kt (hunk 6 above)
Comment (andriy1507) [reply local/src/main/java/com/goryachok/local/LocalDataSource.kt]: Fixed

File: repository/src/main/java/com/goryachok/repository/WeatherRepository.kt (hunk 7 above)
Comment (andriy1507) [reply repository/src/main/java/com/goryachok/repository/WeatherRepository.kt]: Done

File: repository/src/main/java/com/goryachok/repository/WeatherRepository.kt (hunk 8 above)
Comment (andriy1507) [reply repository/src/main/java/com/goryachok/repository/WeatherRepository.kt]: Done

File: repository/src/main/java/com/goryachok/repository/ForecastRepository.kt (hunk 9 above)
Comment (andriy1507) [reply repository/src/main/java/com/goryachok/repository/ForecastRepository.kt]: Done

File: repository/src/main/java/com/goryachok/repository/ForecastRepository.kt (hunk 10 above)
Comment (andriy1507) [reply repository/src/main/java/com/goryachok/repository/ForecastRepository.kt]: Done

File: repository/src/main/java/com/goryachok/repository/WeatherRepository.kt (hunk 11 above)
Comment (andriy1507) [reply repository/src/main/java/com/goryachok/repository/WeatherRepository.kt]: Done

File: repository/src/main/java/com/goryachok/repository/ForecastRepository.kt (hunk 12 above)
Comment (andriy1507) [reply repository/src/main/java/com/goryachok/repository/ForecastRepository.kt]: Done

File: remote/src/main/java/com/goryachok/remote/RemoteDataSource.kt (hunk 15 above)
Comment (andriy1507) [reply remote/src/main/java/com/goryachok/remote/RemoteDataSource.kt]: Done

File: local/src/main/java/com/goryachok/local/model/LocalWeather.kt (hunk 16 above)
Comment (andriy1507) [reply local/src/main/java/com/goryachok/local/model/LocalWeather.kt]: Done

File: local/src/main/java/com/goryachok/local/model/LocalData.kt (hunk 17 above)
Comment (andriy1507) [reply local/src/main/java/com/goryachok/local/model/LocalData.kt]: Done

File: local/src/main/java/com/goryachok/local/LocalDataSource.kt (hunk 18 above)
Comment (andriy1507) [reply local/src/main/java/com/goryachok/local/LocalDataSource.kt:16]: Done

**Comment by andriy1507** [comment]
Maybe better is to make all remote data classes inner?

**Review by EugeneKryvoshey — CHANGES_REQUESTED** [review]

File: remote/src/main/java/com/goryachok/remote/model/Main.kt (hunk 2 above)
Comment (andriy1507) [reply remote/src/main/java/com/goryachok/remote/model/Main.kt]: Done

File: remote/src/main/java/com/goryachok/remote/model/Forecast.kt (hunk 3 above)
Comment (andriy1507) [reply remote/src/main/java/com/goryachok/remote/model/Forecast.kt]: Done

File: remote/src/main/java/com/goryachok/remote/model/Coordinates.kt (hunk 4 above)
Comment (andriy1507) [reply remote/src/main/java/com/goryachok/remote/model/Coordinates.kt]: Done

File: remote/src/main/java/com/goryachok/remote/model/City.kt (hunk 5 above)
Comment (andriy1507) [reply remote/src/main/java/com/goryachok/remote/model/City.kt]: Done

File: remote/src/main/java/com/goryachok/remote/model/Wind.kt (hunk 13 above)
Comment (andriy1507) [reply remote/src/main/java/com/goryachok/remote/model/Wind.kt]: Done

File: remote/src/main/java/com/goryachok/remote/model/Weather.kt (hunk 14 above)
Comment (andriy1507) [reply remote/src/main/java/com/goryachok/remote/model/Weather.kt]: Done

**Review by EugeneKryvoshey — APPROVED** [review]

**Pull request merged**